olive/data/config.py

```diff
@@ -98,30 +98,35 @@ def _update_default_component(self):
     def fill_in_params(self):
         """Fill in the default parameters for each component.
 
-        1. If params_config["component_kwargs"] is not None, use the params_config["component_kwargs"]
-        to update component.params
-        2. if params_config is not None, use the params_config to fill in the params. Overrides the
-        component.params
-        3. if params_config is None, use the default params from the function signature
-        4. if there is already define params under the component, use the params directly
+        For each component, we will do the following steps:
+            1. If params_config["component_kwargs"] is not None, use the params_config["component_kwargs"]
+            to update component.params. Higher priority than the following steps.
+        Loop over the parameters from the function signature of the component:
+            2. If defined in params_config, use it to fill in the params
+            3. Else if already defined in component.params, use it directly
+            4. Else Use the default value from the function signature
+
+        Priority is: component_kwargs > params_config > component.params > default value
         """
         from inspect import signature
 
-        self.params_config = self.params_config or {}
+        self.params_config = deepcopy(self.params_config) or {}
         component_kwargs = self.params_config.pop("component_kwargs", {})
         for k, v in self.components.items():
             component = Registry.get_component(k, v.type)
             # 1. use the params_config["component_kwargs"] to update component.params
```

Comment:
Remove this line to align with the steps you re-write.

Reply:
I think this comment is correct. this line and line 121 together perform step 1 described in line 1. We have to do the update first (line 118) and then skip any potential updates after that line 122)